### ISSUE TITLE
RTMClient.addOutgoingEvent(true, 'presence_sub', { ids: userIds }); never resolves

### ISSUE THREAD

**Issue by travisghansen** [slackapi/node-slack-sdk#706]
### Description

When attempting to send `true` for the `awaitReply` parameter on the `presense_sub` and `presence_query` events the promise never resolves.  This may be by design and it may be happening with other events as well (only tried those 2 beyond the `message` event).

### What type of issue is this? (place an `x` in one of the `[ ]`)
- [X ] bug
- [ ] enhancement (feature request)
- [ ] question
- [ ] documentation related
- [ ] testing related
- [ ] discussion

### Requirements (place an `x` in each of the `[ ]`)
* [ X] I've read and understood the [Contributing guidelines](https://github.com/slackapi/node-slack-sdk/blob/master/.github/contributing.md) and have done my best effort to follow them.
* [ X] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [ X] I've searched for any related issues and avoided creating a duplicate issue.

---

### Bug Report

Filling out the following details about bugs will help us solve your issue sooner.

#### Reproducible in:

`@slack/client` version: 4.9.0

node version: v6.16.0

OS version(s): nodered docker image

#### Steps to reproduce:

1.send event as mentioned in the subject
2. add a `.then()`
3. notice it never fires

#### Expected result:

That eventually `.then()` or `.catch()` would fire.

#### Actual result:

It indefinitely hangs.

#### Attachments:

Logs, screenshots, screencast, sample project, funny gif, etc.

**Comment by aoberoi** [comment]
@travisghansen the [presence_sub](
https://api.slack.com/events/presence_sub) and [presence_query](
https://api.slack.com/events/presence_query) messages don’t receive a reply, so that is the expected behavior. Instead, you should expect to receiver `“presence_change”` events which contain the data you’d need.

The [example in the documentation](
http://slack.dev/node-slack-sdk/rtm_api#subscribing_to_presence_updates) should give you an explanation of how this works. If you’re looming for request/response semantics, using the `WebClient` (as shown) is still an option.

**Comment by travisghansen** [comment]
@aoberoi fair enough, is there a comprehensive (indexed) list of which events are writable and of those which are expected to send responses?  Seems like odd behavior to not respond at all even if the response is "thank you we'll start sending events now and not an actual dataset".

**Comment by aoberoi** [comment]
> is there a comprehensive (indexed) list of which events are writable and of those which are expected to send responses?

There's only 5 (public) writable events on RTM: `message`, `presence_sub`, `presence_query`, `ping`, and `typing`. The comprehensive list [is here](https://api.slack.com/events), but it's not indexed in the way you prefer. The only two that expect responses are `message` and `ping`. You'd have to look at each event type to know this. Instead, we usually guide developers to the [RTM guide](https://api.slack.com/rtm) to understand the capabilities.

The `RTMClient` implementation is meant to make understanding all of that not-strictly necessary. We expose a method for each of the writable events that generally "do the right thing" ([the porcelain](https://stackoverflow.com/a/6976506)), and two lower-level methods (`addOutgoingEvent()` and `send()`) for experts who need very specific behavior (the plumbing).

But wait, there's no method for `presence_query`, right? Yes, that is intentional. The functionality provided by the RTM API's `presence_query` is superseded by the introduction of the `users.getPresence` Web API method - so there's no porcelain method for using it in `RTMClient`.

**TL;DR**: The `RTMClient`'s API is supposed to expose all the functionality and steer users towards the intended usage. We're open to feedback on whether or not its meeting this goal, and to understanding what might have been confusing.

> Seems like odd behavior to not respond at all even if the response is "thank you we'll start sending events now and not an actual dataset".

The main use case for `RTMClient.addOutgoingEvent()` is to send a `message` event that is queued but where we don't need to await the reply. This was possible in earlier versions of the client, and so its still `public` to preserve that functionality.

The truth is that everything over a Websocket (underlying transport for RTM) is fire-and-forget. The server-side of the API has made special considerations to "ack" only the requests where there's client-side behavior that is strictly dependent on it (e.g. `message` events because a Slack client for users would want to render the message as pending right away, but after some time if the server doesn't acknowledge it, it can offer the user a quick way to retry). If we tried to "ack" every message, we'd quickly get into a situation where you're trying to ack an ack, and that seems rather pointless, right? Instead, the paradigm is that you trust the message was delivered, and that all messages you're receiving are correct and reflect what you asked for, except when you can detect that the connection is flaky (which is the reason for `ping` and `pong`).

This means RTM isn't a great transport for anything that can be described as a query. Another way of saying this is RTM is not good for situations where you are looking to group a certain call with a certain result. That's okay, because that's what the Web API is exactly well-suited for.

I hope I didn't bore you with all these details. I'm glad you asked because I feel like this is an area of the platform that's usually a little tricky for people understand.

**Comment by travisghansen** [comment]
@aoberoi That explanation is fantastic!  I've been reading both the Web/RTM code/guides extensively and they've been generally very helpful.

I understand all the implications/etc you've described.  I'm trying to put together a library/integration for `node-red` to `slack`.  In doing so I've made the consumer-side of the API very dynamic (methods and inputs are set by 'consumers' dynamically) which is exactly why I'm using the generics `RTMClient.addOutgoingEvent()` and `WebClient.apiCall()` in my code as the specific methods/options are being generated 'outside' the library and sent in as 'arguments'.

The expectation in flow-based programming is that the output would generally be consistent which is why I was hoping for a consistent response to all methods.  I'll either make up a fake response following the 'normal' response structure or just document it as "don't expect an output if you're invoking x, y, or z methods".
